core/profile.c

```diff
@@ -1096,45 +1103,47 @@ void calculate_deco_information(struct dive *dive, struct divecomputer *dc, stru
 
 				/* We are going to mess up deco state, so store it for later restore */
 				struct deco_state *cache_data = NULL;
-				cache_deco_state(&cache_data);
-				calculate_ndl_tts(dive, entry, gasmix, surface_pressure);
+				cache_deco_state(ds, &cache_data);
+				calculate_ndl_tts(ds, dive, entry, gasmix, surface_pressure);
 				if (decoMode() == VPMB && !in_planner() && i == pi->nr - 1)
 					final_tts = entry->tts_calc;
 				/* Restore "real" deco state for next real time step */
-				restore_deco_state(cache_data, decoMode() == VPMB);
+				restore_deco_state(cache_data, ds, decoMode() == VPMB);
 				free(cache_data);
 			}
 		}
 		if (decoMode() == VPMB && !in_planner()) {
 			int this_deco_time;
-			prev_deco_time = deco_state->deco_time;
+			prev_deco_time = ds->deco_time;
 			// Do we need to update deco_time?
 			if (final_tts > 0)
-				deco_state->deco_time = last_ndl_tts_calc_time + final_tts - time_deep_ceiling;
+				ds->deco_time = last_ndl_tts_calc_time + final_tts - time_deep_ceiling;
 			else if (time_clear_ceiling > 0)
 				/* Consistent with planner, deco_time ends after ascending (20s @9m/min from 3m)
 				 * at end of whole minute after clearing ceiling. The deepest ceiling when planning a dive
 				 * comes typically 10-60s after the end of the bottom time, so add 20s to the calculated
 				 * deco time. */
-				deco_state->deco_time = ROUND_UP(time_clear_ceiling - time_deep_ceiling + 20, 60) + 20;
-			vpmb_next_gradient(deco_state->deco_time, surface_pressure / 1000.0);
+					ds->deco_time = ROUND_UP(time_clear_ceiling - time_deep_ceiling + 20, 60) + 20;
+			vpmb_next_gradient(ds, ds->deco_time, surface_pressure / 1000.0);
 			final_tts = 0;
 			last_ndl_tts_calc_time = 0;
 			first_ceiling = 0;
 			first_iteration = false;
 			count_iteration ++;
-			this_deco_time = deco_state->deco_time;
-			restore_deco_state(cache_data_initial, true);
-			deco_state->deco_time = this_deco_time;
+			this_deco_time = ds->deco_time;
+			restore_deco_state(cache_data_initial, ds, true);
+			ds->deco_time = this_deco_time;
+			restore_deco_state(cache_data_initial, ds, true);
```

Comment:
I guess calling `restore_deco_state(cache_data_initial, ds, true);` here for a second time is not correct.